### DISCUSSION TITLE
solargraph formatting help

### DISCUSSION THREAD

**Discussion by waymondo** [joaotavora/eglot#1137] [Q&A]
I'm trying to get Solargraph to allow me to run `eglot-format-buffer` and I'm not sure the right way to configure it. If I comment out [these lines](https://github.com/joaotavora/eglot/blob/master/eglot.el#L2686-L2687) it does format successfully, otherwise I get the "Server can't format!" message. I'm unfamiliar with the inner workings of LSPs but is it something along the lines of the Solargraph server is not broadcasting it has the appropriate ability in a way that Eglot understands?

FWIW, I had to create a `.dir-locals.el` file for my project to get diagnostics (via Rubocop) to work. Here is the contents of that file:

```
((ruby-mode . ((eglot-workspace-configuration . ((:solargraph :diagnostics t))))))
```

When I add `:formatting` in this same manner, it doesn't have an effect and `(eglot--server-capable :documentFormattingProvider)` still fails. Regardless of any `.dir-local.el` configuration, formatting works if I disable the server capable check.

Any tips would be helpful.

**Comment by doolio** [comment]
I can confirm the same issues. Have you been able to resolve it?

**Reply by waymondo** [reply]
I ended up just disabling the server capable check when `eglot-format` runs.

```emacs-lisp
(defun disable-eglot-format-check (f &rest args)
  (cl-letf (((symbol-function 'eglot--server-capable) #'identity))
    (apply f args)))
(advice-add #'eglot-format :around #'disable-eglot-format-check)
```

Formatting works with this in addition to the `.dir-locals.el` addition above, but I never figured out the core issue.

**Reply by doolio** [reply]
Thanks for this. If I learn anything I'll post back here.
